src/sys/redox/selector.rs

```diff
@@ -0,0 +1,206 @@
+use event::Event;
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fs::File,
+    io::prelude::*,
+    mem,
+    os::unix::io::{AsRawFd, RawFd},
+    slice,
+    sync::{
+        Mutex,
+        atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT}
+    },
+    time::Duration
+};
+use syscall::{self, CLOCK_MONOTONIC, EVENT_READ, EVENT_WRITE};
+use {io, Ready, PollOpt, Token};
+
+const TIMEOUT_TOKEN: Token = Token(::std::usize::MAX - 1);
+
+/// Each Selector has a globally unique(ish) ID associated with it. This ID
+/// gets tracked by `TcpStream`, `TcpListener`, etc... when they are first
+/// registered with the `Selector`. If a type that is previously associatd with
+/// a `Selector` attempts to register itself with a different `Selector`, the
+/// operation will return with an error. This matches windows behavior.
+static NEXT_ID: AtomicUsize = ATOMIC_USIZE_INIT;
+
+#[derive(Debug)]
+pub struct Selector {
+    id: usize,
+    efd: File,
+    tokens: Mutex<BTreeMap<RawFd, BTreeSet<Token>>>
+}
+
+impl Selector {
+    pub fn new() -> io::Result<Selector> {
+        let efd = File::open("event:")?;
+
+        // offset by 1 to avoid choosing 0 as the id of a selector
+        let id = NEXT_ID.fetch_add(1, Ordering::Relaxed) + 1;
+
+        Ok(Selector {
+            id: id,
+            efd: efd,
+            tokens: Mutex::new(BTreeMap::new()),
+        })
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    /// Wait for events from the OS
+    pub fn select(&self, evts: &mut Events, awakener: Token, timeout: Option<Duration>) -> io::Result<bool> {
+        let mut timeout_fd = None;
+        if let Some(timeout) = timeout {
+            let mut file = File::open(format!("time:{}", CLOCK_MONOTONIC))?;
+
+            // TODO: use try_from below when stable
+            if timeout.as_secs() > ::std::i64::MAX as u64 {
+                panic!("too high duration");
+            }
+
+            let mut time = syscall::TimeSpec::default();
+            file.read(&mut time)?;
+
+            //tv_sec += i64::try_from(timeout.as_secs()).expect("too high duration"),
```

Comment:
Remove?

Reply:
When it becomes stable I want to use this code instead.

Reply:
Could you add a note about that? Now it seems like the code is just left in by accident.